### PR TITLE
Add terraform plan check to PR workflow.

### PR DESCRIPTION
**Pull request by rin-skylight**
## Related Issue or Background Info

- Right now, there are instances where SimpleReport bears infrastructure changes that have been made by hand and are not reflected in the existing Terraform state. Until we can more readily enforce the usage of infrastructure as code, we need more visibility around unintended changes that automated deployments can cause.
- This serves as an initial-stage effort for #3138.

## Changes Proposed

- Add a `terraform plan` step that runs against Prod as part of the existing PR checks workflow
- Add a new **Ad-hoc Terraform Plan** workflow that allows TF changes to be planned against lower-level environments.
- Modify the existing PR template to add visibility to the new Terraform enhancements, and hopefully enforce proper review of the generated plans.
- Update the existing app instance number for PROD to reflect what exists today.

### CODE REVIEW

File: ops/prod/api.tf [hunk 1]
```diff
@@ -3,7 +3,7 @@ module "simple_report_api" {
   name   = "${local.name}-api"
   env    = local.env
 
-  instance_count = 3
+  instance_count = 4
```
Comment (jdorothy) [inline ops/prod/api.tf:6]: Can we make this a separate commit? It's unrelated to adding `terraform plan`.

File: .github/workflows/terraformPlan.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+name: Ad-hoc Terraform Plan
```
Comment (jdorothy) [inline .github/workflows/terraformPlan.yml:1]: I'm not entirely sure I see the need for this workflow when you can just run the same locally. I wouldn't expect to run unit tests via a workflow.

**Review by jdorothy — CHANGES_REQUESTED** [review]

File: ops/prod/api.tf (hunk 1 above)
Comment (rin-skylight) [reply ops/prod/api.tf]: Absolutely! Done. I'll put that right-size in a separate PR.

File: .github/workflows/terraformPlan.yml (hunk 2 above)
Comment (rin-skylight) [reply .github/workflows/terraformPlan.yml:1]: That's a great callout, and I absolutely understand where you are coming from. This was primarily created to allow for checking effects against different environments, since the `terraform_checks` additions only operate against PROD. Plus, it allows members who don't have direct Azure access to see the effects of code changes.

I'm perfectly fine removing this, though! This addition doesn't serve as much of a safety purpose as the other changes in this PR do.

File: ops/Makefile [hunk 3]
```diff
@@ -46,6 +46,10 @@ init-%: .valid-env-%
 	terraform -chdir=$*/persistent init
 	terraform -chdir=$* init
 
+plan-%: .valid-env-% api.tfvars
+	terraform -chdir=$*/persistent plan -lock=false
```
Comment (jdorothy) [inline ops/Makefile:50]: We shouldn't need to avoid locking here (at least I don't think), because multiple GH Actions using the same workflow will queue themselves instead of running concurrently.

**Review by jdorothy — COMMENTED** [review]
One minor comment - LGTM otherwise!

File: .github/workflows/terraformPlan.yml (hunk 2 above)
Comment (jdorothy) [reply .github/workflows/terraformPlan.yml:1]: That's a really good point about being able to run `terraform plan` without Azure access. I'm fine with keeping this in that case.

**Review by jdorothy — APPROVED** [review]
One minor comment re: locking but LGTM!

File: ops/Makefile (hunk 3 above)
Comment (rin-skylight) [reply ops/Makefile:50]: That's a good callout! The reason why I added this was because I experienced a situation where one of the plan actions caused by the GH Action run locked the state, then was cancelled, which required my having to ask for someone to go into Azure and manually execute a forcible unlock. (Thank you again for helping me with that, by the way!) I disabled locking here to prevent that possibility. I'm happy to remove it if there's a compelling argument for doing so, though!

Comment (jdorothy) [reply ops/Makefile:50]: I think that's fixing the symptoms rather than the cause, which is forcibly interrupting Terraform while it's holding the lock. We want to Terraform to fail if it can't hold a lock, rather than avoiding asking for one, so we can guarantee state correctness. I'd much rather manually remove the lock every once in a while.

Comment (rin-skylight) [reply ops/Makefile:50]: Understood! I totally see where you're coming from. Just to make fully sure, as a disclosure: if we run into this situation, deployments to any/all environments will be blocked until someone manually intervenes. There's always the possibility that GH kills off one of its own runners, or some other blip with a system causes this to happen without a human having caused it. As we move toward greater resiliency and additional automation, forcing a lock here introduces a point of failure/hand-touchy. Ideally, we will want to move toward limiting who has Azure SU access as we move forward and/or transition this as it becomes more mature, so that requirement could add a bit more of a burden to operations personnel.

Again, totally fine doing this, and will happily push a commit to do so later today! I want to make sure we're on the same page before I do.

Comment (jdorothy) [reply ops/Makefile:50]: I agree with everything you've said. That said, I think I've seen a non-human caused Terraform termination maybe twice ever, so in my mind the likelihood of this being a recurring problem is very low. Additionally, when there's a lock failure, there's an RCA we're going to want to do to make sure this isn't a real problem, and a held lock prevents someone else from updating the infrastructure that may or may not be in a coalesced state in the meantime.

**Review by emmastephenson — DISMISSED** [review]
LGTM!

**Review by alismx — APPROVED** [review]
Love it!

File: ops/Makefile (hunk 3 above)
Comment (rin-skylight) [reply ops/Makefile]: Ah, I hadn't considered the RCA need! Thanks for that extra bit of context, and for this really constructive discussion. I've re-enabled the locking on this, per your comments.

**Review by jdorothy — APPROVED** [review]

**Review by alismx — APPROVED** [review]
Thanks for adding this.
